internal_ws/ykshim/src/test_api.rs

```diff
@@ -2,135 +2,146 @@
-    *ret_tyid = tyid;
-}
-
-/// Creates a TraceCompiler with default settings.
-#[no_mangle]
-unsafe extern "C" fn __ykshimtest_tracecompiler_default() -> *mut TraceCompiler {
-    let tc = Box::new(TraceCompiler::new(Default::default(), Default::default()));
-    Box::into_raw(tc)
-}
-
-/// Drop a TraceCompiler.
-#[no_mangle]
-unsafe extern "C" fn __ykshimtest_tracecompiler_drop(comp: *mut c_void) {
-    Box::from_raw(comp as *mut TraceCompiler);
-}
-
-/// Inserts a local declaration of the specified TypeId into a TraceCompiler.
-#[no_mangle]
-unsafe extern "C" fn __ykshimtest_tracecompiler_insert_decl(
-    tc: *mut c_void,
-    local: Local,
-    local_ty: TypeId,
-    referenced: bool,
-) {
-    let tc = &mut *(tc as *mut TraceCompiler);
-    tc.local_decls.insert(
-        local,
-        LocalDecl {
-            ty: local_ty,
-            referenced,
-        },
-    );
-}
-
-/// Returns a string describing the register allocation of the specified local.
-#[no_mangle]
-unsafe extern "C" fn __ykshimtest_tracecompiler_local_to_location_str(
-    tc: *mut c_void,
-    local: Local,
-) -> *mut c_char {
-    let tc = &mut *(tc as *mut TraceCompiler);
-    let rstr = format!("{:?}", tc.local_to_location(local));
-    CString::new(rstr.as_str()).unwrap().into_raw()
-}
-
-/// Inform a TraceCompiler's register allocator that a local variable is dead.
-#[no_mangle]
-unsafe extern "C" fn __ykshimtest_tracecompiler_local_dead(tc: *mut TraceCompiler, local: Local) {
-    let tc = &mut *tc;
-    tc.local_dead(&local).unwrap();
-}
-
-/// Find a symbol's address in the current memory image. Returns NULL if it can't be found.
-#[no_mangle]
-unsafe extern "C" fn __ykshimtest_find_symbol(sym: *const c_char) -> *mut c_void {
-    find_symbol(CStr::from_ptr(sym).to_str().unwrap()).unwrap_or_else(|_| ptr::null_mut())
-}
-
-/// Interpret a SIR body with the specified interpreter context.
-#[no_mangle]
-unsafe extern "C" fn __ykshimtest_interpret_body(body_name: *const c_char, ctx: *mut u8) {
-    let fname = CStr::from_ptr(body_name).to_str().unwrap().to_string();
-    let mut si = StopgapInterpreter::from_symbol(fname);
-    si.set_interp_ctx(ctx);
-    si.interpret();
-}
-
-/// Returns the size of the register allocators register pool.
-#[no_mangle]
-unsafe extern "C" fn __ykshimtest_reg_pool_size() -> usize {
-    REG_POOL.len()
-}
-
-/// Consumes and compiles the given TIR trace to native code, returning an opaque pointer to the
-/// compiled trace.
-#[no_mangle]
-fn __ykshimtest_compile_tir_trace(tir_trace: *mut TirTrace) -> *mut CompiledTrace {
-    let tir_trace = unsafe { Box::from_raw(tir_trace) };
-    let compiled_trace = ykcompile::compile_trace(*tir_trace);
-    Box::into_raw(Box::new(compiled_trace))
+// To avoid marking each individual import/struct/function/etc. as "#[cfg(feature="testing")]", we
+// stuff everything inside this conditionally compiled "testing" module.
+#[cfg(feature = "testing")]
```

Comment:
why don't we do this annotation one level up in `lib.rs`?

```
#[cfg(testing)]
mod test_api;
```

Reply:
AFAICS `cargo` still builds files even if they're not `mod`ed, so you end up with compilation errors.

Reply:
No, it doesn't. There were proposals to make `mod` implicit in the 2018 edition, but those were not accepted.